jetty-client/src/main/java/org/eclipse/jetty/client/ProxyProtocolClientConnectionFactory.java

```diff
@@ -0,0 +1,561 @@
+    }
+
+    public ClientConnectionFactory getClientConnectionFactory()
+    {
+        return factory;
+    }
+
+    @Override
+    public org.eclipse.jetty.io.Connection newConnection(EndPoint endPoint, Map<String, Object> context)
+    {
+        ProxyProtocolConnection connection = newProxyProtocolConnection(endPoint, context);
+        return customize(connection, context);
+    }
+
+    protected abstract ProxyProtocolConnection newProxyProtocolConnection(EndPoint endPoint, Map<String, Object> context);
+
+    private abstract static class ProxyProtocolConnection extends AbstractConnection implements Callback
+    {
+        protected static final Logger LOG = Log.getLogger(ProxyProtocolConnection.class);
+
+        private final ClientConnectionFactory factory;
+        private final Map<String, Object> context;
+
+        private ProxyProtocolConnection(EndPoint endPoint, Executor executor, ClientConnectionFactory factory, Map<String, Object> context)
+        {
+            super(endPoint, executor);
+            this.factory = factory;
+            this.context = context;
+        }
+
+        @Override
+        public void onOpen()
+        {
+            super.onOpen();
+            writePROXYBytes(getEndPoint(), this);
+        }
+
+        protected abstract void writePROXYBytes(EndPoint endPoint, Callback callback);
+
+        @Override
+        public void succeeded()
+        {
+            try
+            {
+                EndPoint endPoint = getEndPoint();
+                Connection connection = factory.newConnection(endPoint, context);
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Written PROXY line, upgrading to {}", connection);
+                endPoint.upgrade(connection);
+            }
+            catch (Throwable x)
+            {
+                failed(x);
+            }
+        }
+
+        @Override
+        public void failed(Throwable x)
+        {
+            close();
+            Promise<?> promise = (Promise<?>)context.get(HttpClientTransport.HTTP_CONNECTION_PROMISE_CONTEXT_KEY);
+            promise.failed(x);
+        }
+
+        @Override
+        public InvocationType getInvocationType()
+        {
+            return InvocationType.NON_BLOCKING;
+        }
+
+        @Override
+        public void onFillable()
+        {
+        }
+    }
+
+    private static class ProxyProtocolConnectionV1 extends ProxyProtocolConnection
+    {
+        private final V1.Tag tag;
+
+        public ProxyProtocolConnectionV1(EndPoint endPoint, Executor executor, ClientConnectionFactory factory, Map<String, Object> context, V1.Tag tag)
+        {
+            super(endPoint, executor, factory, context);
+            this.tag = tag;
+        }
+
+        @Override
+        protected void writePROXYBytes(EndPoint endPoint, Callback callback)
+        {
+            try
+            {
+                InetSocketAddress localAddress = endPoint.getLocalAddress();
```

Comment:
Do all our endpoints return non null InetAddresses?   What if this is applied to a UDP endpoint?